Rebuild widgets screen (and equivalent in Customizer) to use details and summary tags

KTS915 opened the pull request:
This PR follows those for the menu screen, dashboard widgets, and post/page metaboxes in replacing widgets on the admin widgets screen with HTML5 disclosure widgets, which utilize the `details` and `summary` tags.

## Motivation and context
Addresses Issue #224.
Improves accessibility and reduces the use of JavaScript.

## How has this been tested?
On my localhost test site

## Screenshots
### After
![Screenshot at 2023-10-07 00-46-42](https://github.com/ClassicPress/ClassicPress-v2/assets/4127662/c30c4291-5f5e-4cbd-ad5c-3566763796e1)
![Screenshot at 2023-10-07 09-45-45](https://github.com/ClassicPress/ClassicPress-v2/assets/4127662/bdf56666-97e0-4e10-b004-a7a6ea0755e4)

## Types of changes
- New feature

KTS915 committed:
Update widgets.php to use `details` and `summary` tags

This significantly improves accessibility while also updating the admin to use HTML5 and less JavaScript.

KTS915 committed:
Update widgets-form.php

KTS915 committed:
Update class-wp-customize-widgets.php to work with `details` and `summary` elements

KTS915 committed:
Update customize-widgets.js to work with more semantic markup

KTS915 committed:
Update media-widgets.js

Enables all the media widgets (image, audio, video, etc) to work with more semantic markup and the `details` and `summary` elements.

KTS915 committed:
Update text-widgets.js to work with more semantic markup

KTS915 committed:
Update custom-html-widgets.js to work with more semantic markup

KTS915 committed:
Update widgets.js to work with `details` and `summary` elements and more semantic  markup

KTS915 committed:
Update widgets.css so that changes to markup cause minimal visual differences.

KTS915 committed:
Update widgets.js to ensure chooser reamins available after dragging a widget

KTS915 committed:
Update widgets.js to ensure that `inert` attribute on widget contents is removed after dragging to sidebar.

KTS915 committed:
Update widgets.js to increment each new added widget's ID.

KTS915 committed:
Update widgets.js to ensure widget opens when added to sidebar.

KTS915 committed:
Update widgets.css to ensure available widgets sidebar description is visible.

KTS915 committed:
Update widgets.php to improve semantics of receiving sidebar

KTS915 committed:
Update widgets.js to have receiving sidebars open and close when widget dragged over or off them.

KTS915 committed:
Merge branch 'ClassicPress:develop' into better-widgets

KTS915 committed:
Update customize-controls.css

KTS915 committed:
Update customize-widgets.css

KTS915 committed:
Update customize-widgets.css to ensure summary marker does not show in Safari

KTS915 committed:
Update widgets.css to enable custom `summary` marker

KTS915 committed:
Update widgets.php to reinstate `<h3>` header after feedback on dashboard widgets PR

KTS915 committed:
Update widgets.css to create custom markers for lists of widget

KTS915 committed:
Update widgets.php to reinstate `<h2>` headers

KTS915 committed:
Update widgets-form.php to reinstate `<h2>` headers

KTS915 committed:
Update widgets.css to account for reinstatement of `<h2>` headers

KTS915 committed:
Update widgets.css to maintain marker spacing

KTS915 committed:
Update widgets.css

mattyrob committed:
PHP Coding Standards

mattyrob commented:
@KTS915 - I've been looking into the test failures and trying to fix what I can.

There are 2 failing PHPUnit tests, once is a simple fix, the other links to the change in `src/wp-admin/includes/widgets.php` from:
`$params[0]['before_widget'] = "<div id='widget-{$i}_{$id}' class='widget'$hidden>";`
To:
`$params[0]['before_widget'] = '<'. $tag . ' id="widget-' . $i . '_' . $id . '" class="widget open">';`

The test now fails for 2 reasons, a regular expression is looking for an `<li>` tag on a new line with a class of `widget` contains in single quotes, and it's the only class.

The current code changes the tag, and also the quote type from single to double quotes and it also adds the `open` class name. Would it still work without the open class name added and with the line as follows:
`$params[0]['before_widget'] = "<{$tag} id='widget-{$i}_{$id}' class='widget'>";`

KTS915 commented:
> @KTS915 - I've been looking into the test failures and trying to fix what I can.
> 
> There are 2 failing PHPUnit tests, once is a simple fix, the other links to the change in `src/wp-admin/includes/widgets.php` from: `$params[0]['before_widget'] = "<div id='widget-{$i}_{$id}' class='widget'$hidden>";` To: `$params[0]['before_widget'] = '<'. $tag . ' id="widget-' . $i . '_' . $id . '" class="widget open">';`
> 
> The test now fails for 2 reasons, a regular expression is looking for an `<li>` tag on a new line with a class of `widget` contains in single quotes, and it's the only class.
> 
> The current code changes the tag, and also the quote type from single to double quotes and it also adds the `open` class name. Would it still work without the open class name added and with the line as follows: `$params[0]['before_widget'] = "<{$tag} id='widget-{$i}_{$id}' class='widget'>";`

Thanks very much for doing all that, @mattyrob!

I remember I added the class `open` to stop the widget contents being hidden when a widget was opened. But it seems that later changes I made have rendered that unnecessary because, yes, your suggestion works perfectly!

mattyrob committed:
PHPUnit fixes

mattyrob committed:
JSHint fixes

mattyrob commented:
@KTS915 - could you just double check that the JavaScript coding standards changes haven't broken anything in usage and we can consider this for merge later this week.

KTS915 committed:
Update widgets.js to address error message in console

KTS915 committed:
Update media-widgets.js to make widget open when added to sidebar

KTS915 commented:
@mattyrob I've just pushed a commit to eradicate an error message in the console. Other than that, I can't see any problems with your fixes (and I doubt that one was caused by anything you did anyway).

However, I also found that the PHP change to using `$params[0]['before_widget'] = "<{$tag} id='widget-{$i}_{$id}' class='widget'>";` in `widgets.php` _did_ affect _some_ widgets (the ones that use media, such as Image, Video, and Audio). Obviously, I just didn't test the right ones before. The issue is, indeed, the lack of the class of `open`.

But I've now found a better way of addressing that by checking for the `open` attribute instead and have pushed that fix too. So now I think we're good to go.

xxsimoxx reviewed (changes requested):
~Having problems in safari. Widget section is not looking good, and in the customizer the toggles are not working.~ 
<img width="337" alt="Schermata 2023-10-10 alle 10 26 39" src="https://github.com/ClassicPress/ClassicPress-v2/assets/29772709/f8e420af-b395-4a15-bd8f-e98a60091a8e">


Looks good in Chrome and Firefox.
Even a better UX for me:
Before:
<img width="298" alt="Schermata 2023-10-10 alle 10 24 01" src="https://github.com/ClassicPress/ClassicPress-v2/assets/29772709/5963317b-3d73-4eeb-b9c4-d1a4182431fc">
After:

<img width="289" alt="Schermata 2023-10-10 alle 10 33 50" src="https://github.com/ClassicPress/ClassicPress-v2/assets/29772709/f2bffcfd-0307-436c-a580-6b254c7a2a88">

KTS915 committed:
Update widgets.css to use native disclosure widget marker, but enlarged

I have just read an article by Scott O'Hara (https://www.scottohara.me/blog/2022/09/12/details-summary.html), where he says that removing the native marker for the disclosure widget can cause accessibility problems because some assistive technology looks to the marker instead of the status of the `open` attribute to know whether the widget is open or not! Yes, it's completely crazy!

So this commit goes back to using the native marker but makes it bigger. Which, really, I suppose is what I should have done in the first place. I will submit a PR to make a similar change for the dashboard widgets, post metaboxes, and menu items that we've already included.

KTS915 commented:
@xxsimoxx I haven't got a way yet of testing on Safari directly, but I have installed GNOME Web, which uses the same rendering engine (and uses the same markup for the disclosure widget marker). I found much the same as you initially but, when I did a hard refresh, it worked properly.

If a hard refresh doesn't work for you, I'll set up a site outside my desktop and test with Safari online.

xxsimoxx commented:
@KTS915 a (command line) cache reset made the trick on safari.

viktorix reviewed (approved):
Good on Windows.
